src/Databases/DatabaseReplicated.cpp

```diff
@@ -0,0 +1,288 @@
+    current_zookeeper->createIfNotExists(zookeeper_path + "/log", String());
+    current_zookeeper->createIfNotExists(zookeeper_path + "/snapshots", String());
+    current_zookeeper->createIfNotExists(zookeeper_path + "/replicas", String());
+}
+
+void DatabaseReplicated::RemoveOutdatedSnapshotsAndLog() {
+    // This method removes all snapshots and logged queries
+    // that no longer will be in use by current replicas or
+    // new coming ones.
+    // Each registered replica has its state in ZooKeeper.
+    // Therefore, snapshots and logged queries that are less
+    // than a least advanced replica are removed.
+    // It does not interfere with a new coming replica
+    // metadata loading from snapshot
+    // because the replica will use the latest snapshot available
+    // and this snapshot will set the last executed log query
+    // to a greater one than the least advanced current replica.
+    current_zookeeper = getZooKeeper();
+    Strings replica_states = current_zookeeper->getChildren(zookeeper_path + "/replicas");
+    auto least_advanced = std::min_element(replica_states.begin(), replica_states.end());
+    Strings snapshots = current_zookeeper->getChildren(zookeeper_path + "/snapshots");
+    
+    if (snapshots.size() < 2) {
+        return;
+    }
+
+    std::sort(snapshots.begin(), snapshots.end());
+    auto still_useful = std::lower_bound(snapshots.begin(), snapshots.end(), *least_advanced);
+    snapshots.erase(still_useful, snapshots.end());
+    for (const String & snapshot : snapshots) {
+        current_zookeeper->tryRemoveRecursive(zookeeper_path + "/snapshots/" + snapshot);
+    }
+
+    Strings log_entry_names = current_zookeeper->getChildren(zookeeper_path + "/log");
+    std::sort(log_entry_names.begin(), log_entry_names.end());
+    auto still_useful_log = std::upper_bound(log_entry_names.begin(), log_entry_names.end(), *still_useful);
+    log_entry_names.erase(still_useful_log, log_entry_names.end());
+    for (const String & log_entry_name : log_entry_names) {
+        String log_entry_path = zookeeper_path + "/log/" + log_entry_name;
+        current_zookeeper->tryRemove(log_entry_path);
+    }
+}
+
+void DatabaseReplicated::runBackgroundLogExecutor() {
+    if (last_executed_log_entry == "") {
+        loadMetadataFromSnapshot();
+    }
+
+    current_zookeeper = getZooKeeper();
+    Strings log_entry_names = current_zookeeper->getChildren(zookeeper_path + "/log");
+
+    std::sort(log_entry_names.begin(), log_entry_names.end());
+    auto newest_entry_it = std::upper_bound(log_entry_names.begin(), log_entry_names.end(), last_executed_log_entry);
+
+    log_entry_names.erase(log_entry_names.begin(), newest_entry_it);
+
+    for (const String & log_entry_name : log_entry_names) {
+        String log_entry_path = zookeeper_path + "/log/" + log_entry_name;
+        executeFromZK(log_entry_path);
+        last_executed_log_entry = log_entry_name;
+        writeLastExecutedToDiskAndZK();
```

Comment:
What if query has failed?

Reply:
it will be logged to `LOG_ERROR` and marked by replica as executed (`last_executed_log_entry = log_entry_name`).